packages/contracts/src/contracts/current/protocol/Exchange/MixinMatchOrders.sol

```diff
@@ -0,0 +1,285 @@
+            ) = calculateFillAmounts(
+                rightOrder,
+                rightOrderStatus,
+                rightOrderFilledAmount,
+                rightTakerAssetAmountRemaining);
+            if(status != uint8(Status.SUCCESS)) {
+                return (status, matchedFillOrderAmounts);
+            }
+
+            // The left order just spent <rightTakerAssetAmountRemaining> of their maker asset to fill the right order.
+            // The amount left gets in return is:
+            //    <rightOrderAmountBought> * <rightCostPerUnitSold>
+            //   (let Y = <matchedFillOrderAmounts.right.takerAssetFilledAmount>; let X = matchedFillOrderAmounts.right.makerAssetFilledAmount)
+            // = Y * X / Y
+            // = X = <matchedFillOrderAmounts.right.makerAssetFilledAmount>
+            //
+            // Sanity check: that the amount transferred by the right order does not exceed the amount required to fill the left order.
+            assert(matchedFillOrderAmounts.right.makerAssetFilledAmount <= leftTakerAssetAmountRemaining);
+            (   status,
+                matchedFillOrderAmounts.left
+            ) = calculateFillAmounts(
+                leftOrder,
+                leftOrderStatus,
+                leftOrderFilledAmount,
+                matchedFillOrderAmounts.right.makerAssetFilledAmount);
+            if(status != uint8(Status.SUCCESS)) {
+                return (status, matchedFillOrderAmounts);
+            }
+
+            // Sanity check: the amount sent from the right order must equal the amount received by the left order.
+            assert(matchedFillOrderAmounts.right.makerAssetFilledAmount == matchedFillOrderAmounts.left.takerAssetFilledAmount);
+        }
+
+        return (status, matchedFillOrderAmounts);
+    }
+
+    /// @dev Match two complementary orders that have a positive spread.
+    ///      Each order is filled at their respective price point. However, the calculations are
+    ///      carried out as though the orders are both being filled at the right order's price point.
+    ///      The profit made by the left order goes to the taker (who matched the two orders).
+    /// @param leftOrder First order to match.
+    /// @param rightOrder Second order to match.
+    /// @param leftSignature Proof that order was created by the left maker.
+    /// @param rightSignature Proof that order was created by the right maker.
+    /// @return leftFillResults Amounts filled and fees paid by maker and taker of left order.
+    /// @return leftFillResults Amounts filled and fees paid by maker and taker of right order.
+    function matchOrders(
+        Order memory leftOrder,
+        Order memory rightOrder,
+        bytes leftSignature,
+        bytes rightSignature)
+        public
+        returns (
+            MatchedOrderFillAmounts memory matchedFillOrderAmounts)
+    {
+        // Get left status
+        OrderInfo memory leftOrderInfo;
```

Comment:
Is this a struct because of the stack variable limitations?

Reply:
Yeah =\ ... I left a note on the struct definition.